Verify ENS in the background

cammellos opened the pull request:
Currently ENS are verified explicitly by status-react, this is not ideal
as if that fails it will have to be explicilty retried in status-react.
This commits changes that behavior so that ENS are verified in a loop
and updated if new messages are received.

status: ready

cammellos committed:
Verify ENS in the background

Currently ENS are verified explicitly by status-react, this is not ideal
as if that fails it will have to be explicilty retried in status-react.
This commits changes that behavior so that ENS are verified in a loop
and updated if new messages are received.

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (6)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 7eefde74 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1824/1/) | 2020-01-31 12:14:00 | ~49 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200131-121310-7eefde.zip) |
| :heavy_check_mark: | 7eefde74 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1824/1/) | 2020-01-31 12:18:20 | ~5 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200131-121309-7eefde.zip) |
| :heavy_check_mark: | 7eefde74 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1824/1/) | 2020-01-31 12:21:25 | ~8 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200131-121308-7eefde.aar) |
| | | | | | | |
| :heavy_check_mark: | 9c020474 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1824/2/) | 2020-01-31 14:44:38 | ~46 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200131-144352-9c0204.zip) |
| :heavy_check_mark: | 9c020474 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1824/2/) | 2020-01-31 14:49:09 | ~5 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200131-144351-9c0204.zip) |
| :heavy_check_mark: | 9c020474 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1824/2/) | 2020-01-31 14:52:34 | ~8 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200131-144350-9c0204.aar) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | fb2f5aba | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1824/3/) | 2020-02-04 10:57:05 | ~39 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200204-105625-fb2f5a.zip) |
| :heavy_check_mark: | fb2f5aba | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1824/3/) | 2020-02-04 10:59:41 | ~3 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200204-105625-fb2f5a.zip) |
| :heavy_check_mark: | fb2f5aba | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1824/3/) | 2020-02-04 11:09:38 | ~13 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200204-105623-fb2f5a.aar) |
| | | | | | | |
| :heavy_check_mark: | e153c601 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1824/4/) | 2020-02-04 13:54:17 | ~45 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200204-135331-e153c6.zip) |
| :heavy_check_mark: | e153c601 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1824/4/) | 2020-02-04 13:56:25 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200204-135330-e153c6.zip) |
| :heavy_check_mark: | e153c601 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1824/4/) | 2020-02-04 13:59:34 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200204-135329-e153c6.aar) |

flexsurfer commented:
please do not merge until  https://github.com/status-im/status-go/pull/1825  will be merged

jakubgs commented on protocol/ens.go:
```diff
@@ -0,0 +1,57 @@
+		return false
+	}
+
+	if now < c.ENSVerifiedAt+ENSBackoffTimeMs*c.ENSVerificationRetries*uint64(math.Exp2(float64(c.ENSVerificationRetries))) {
```
A mouthful like this would be nice to be replaced with a `verifiedRecentlyEnough()` helper, something like that. Possibly with a comment about what the `math.Exp2()` is giving us here.

jakubgs commented on protocol/message_handler.go:
```diff
@@ -337,6 +337,17 @@ func (m *MessageHandler) HandleChatMessage(state *ReceivedMessageState) error {
 	state.AllChats[chat.ID] = chat
 
+	contact := state.CurrentMessageState.Contact
+	if overrideENSName(contact, receivedMessage.EnsName, receivedMessage.Clock) {
```
This would probably make more sense as `hasENSNameChanged()`, and maybe add comment like `This should trigger ENS re-verification in VerifyENSNames()`.

jakubgs commented on params/config.go:
```diff
@@ -527,6 +527,9 @@ type ShhextConfig struct {
 	VerifyENSURL string
 
+	// VerifyENSContractAddress is the address of the contract used to verify ENS
+	VerifyENSContractAddress string
```
Maybe add comment explaining that `No default is provided and if not set ENS resolution is disabled` or similar.

jakubgs reviewed (commented):
Overall looks good, just few small comments. My main one would be:

Why are we adding ENS stuff into `protocol` package?
Wouldn't it make sense to make a separate `ens` package?
Or at least `protocol/ens`?

jakubgs commented on protocol/message_handler.go:
```diff
@@ -337,6 +337,17 @@ func (m *MessageHandler) HandleChatMessage(state *ReceivedMessageState) error {
+		contact.ENSVerifiedAt = 0
+		contact.ENSVerified = false
+		contact.ENSVerificationRetries = 0
+		contact.LastENSClockValue = receivedMessage.Clock
```
What do you think about adding something like:
```go
func (*c Contact) ResetENSVerification(clockValue uint64) {
	contact.ENSVerifiedAt = 0
	contact.ENSVerified = false
	contact.ENSVerificationRetries = 0
	contact.LastENSClockValue = clockValue
}
```
In `protocol/contact.go`.

cammellos commented:
Thanks for the review, addressed the feedback:

> Why are we adding ENS stuff into protocol package?

The protocol naming is not very descriptive of the responsibilities, basically the `protocol/` package takes care of the handling and state management of the app, and this happens in reaction to a protocol event (receiving a message from a user).

I would move it to `protocol/ens.go`, but currently there would be some circular dependencies as most of the data structures are top level `Contact` for example, and `ENS` depends on `Contact`, while `Messenger` depends on both `Contact` and `ENS`, so we should move first `Contact` to his own namespace and then have `ENS` and  `Messenger` both depend on it, but I'd rather handle this once adam is back, as he is the one who has a better feel on how things should be organized, if that's ok.
